src/main/cpp/jni/ir_sahab_regexmatcher_RegexMatcher.cc

```diff
@@ -22,71 +24,87 @@ const char* java_pattern_preparation_exception_path = "ir/sahab/regexmatcher/exc
 //       created instances. An instance is created by calling newInstance() and it is destroyed by calling close().
 unsigned int num_created_instances = 0;
 std::map<unsigned int, std::unique_ptr<HyperscanWrapper> > instances;
+std::mutex instances_map_mutex;
 
-static HyperscanWrapper* GetHyperscanInstance(JNIEnv* jenv, jlong handle) {
+static void ThrowJavaException(JNIEnv* jenv, const char* java_error_class_path, std::string message) {
+    jclass clazz = jenv->FindClass(java_error_class_path);
+    if (jenv->ThrowNew(clazz, message.c_str()) < 0) {
+        fprintf(stderr, "Failed to throw exception: %s.", message.c_str());
+        std::exit(EXIT_FAILURE);
+    }
+}
+
+static HyperscanWrapper* GetHyperscanInstance(JNIEnv* jenv, jlong jinstance_id) {
     HyperscanWrapper* instance = nullptr;
+    auto instance_id = static_cast<unsigned int>(jinstance_id);
     try {
-        instance = instances.at(static_cast<unsigned int>(handle)).get();
+        instances_map_mutex.lock();
+        instance = instances.at(instance_id).get();
     } catch (std::out_of_range& e) {
-        jclass clazz = jenv->FindClass(java_assertion_error_path);
-        std::string msg = "Invalid instance handle: handle = ";
-        msg += handle;
-        if (jenv->ThrowNew(clazz, msg.c_str()) < 0) {
-            fprintf(stderr, "Failed to throw exception: %s.", msg.c_str());
-            std::exit(EXIT_FAILURE);
-        }
+        ThrowJavaException(jenv, java_assertion_error_path, "Either instance closed or not valid: Instance ID = "
+            + std::to_string(instance_id));
     }
+    instances_map_mutex.unlock();
     return instance;
 }
 
 JNIEXPORT jlong JNICALL Java_ir_sahab_regexmatcher_RegexMatcher_newInstance(
         JNIEnv* jenv, jobject jobj) {
-    ++num_created_instances;
+    instances_map_mutex.lock();
+    num_created_instances++;
     instances[num_created_instances] = std::unique_ptr<HyperscanWrapper>(new HyperscanWrapper());
+    instances_map_mutex.unlock();
     return num_created_instances;
 }
 
 JNIEXPORT void JNICALL Java_ir_sahab_regexmatcher_RegexMatcher_close(
-        JNIEnv* jenv, jobject jobj, jlong jinstance_handle) {
-    instances.erase(static_cast<unsigned int>(jinstance_handle));
+        JNIEnv* jenv, jobject jobj, jlong jinstance_id) {
+    instances_map_mutex.lock();
+    instances.erase(static_cast<unsigned int>(jinstance_id));
+    instances_map_mutex.unlock();
 }
 
 JNIEXPORT void JNICALL Java_ir_sahab_regexmatcher_RegexMatcher_addPattern(
-        JNIEnv* jenv, jobject jobj, jlong jinstance_handle, jlong jpattern_id, jstring jpattern,
+        JNIEnv* jenv, jobject jobj, jlong jinstance_id, jlong jpattern_id, jstring jpattern,
```

Comment:
I think it is good to avoid adding patterns with negative ID. (throw illegal argument exception)